conformance/tests/gateway-observed-generation-bump.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, GatewayObservedGenerationBump)
+}
+
+var GatewayObservedGenerationBump = suite.ConformanceTest{
+	ShortName:   "GatewayObservedGenerationBump",
+	Description: "A Gateway in the gateway-conformance-infra namespace should update the observedGeneration in all of it's Status.Conditions after an update to the spec",
+	Manifests:   []string{"tests/gateway-observed-generation-bump.yaml"},
+	Test: func(t *testing.T, s *suite.ConformanceTestSuite) {
+
+		gwNN := types.NamespacedName{Name: "observed-generation-bump", Namespace: "gateway-conformance-infra"}
+
+		t.Run("observedGeneration should increment", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			namespaces := []string{"gateway-conformance-infra"}
+			kubernetes.NamespacesMustBeAccepted(t, s.Client, s.TimeoutConfig, namespaces)
+
+			existing := &v1beta1.Gateway{}
+			err := s.Client.Get(ctx, gwNN, existing)
+			require.NoErrorf(t, err, "error getting Gateway: %v", err)
+
+			// Sanity check
+			if kubernetes.GatewayConditionsHaveLatestObservedGeneration(existing) {
+				t.Fatal("Not all the condition's observedGeneration were updated")
+			}
+
+			all := v1beta1.NamespacesFromAll
+
+			// mutate the Gateway Spec
+			existing.Spec.Listeners = append(existing.Spec.Listeners, v1beta1.Listener{
+				Name:     "alternate",
+				Port:     8080,
```

Comment:
at the moment, contour will not mark this update as accepted, we don't support a port other than 80 or 443

Reply:
can we do something trivial here like you've done with gw class, or add an annotation etc.?

Reply:
or maybe change an existing listeners allowedroutes section?

Reply:
> we don't support a port other than 80 or 443

This restriction is odd - the API doesn't specify any constraints on the port numbers

Reply:
well I may be misspeaking a little, but at the very least Contour doesn't let you configure multiple HTTP or HTTPS Listeners, only one port is allowed for each at the moment

Reply:
Created an issue to move this discussion off the PR - https://github.com/kubernetes-sigs/gateway-api/issues/1607

Reply:
this test is passing now on Contour since observed generation is set correctly, the status we get is:

```
status:
  addresses:
  - type: IPAddress
    value: 172.24.255.203
  conditions:
  - lastTransitionTime: "2022-12-16T19:15:41Z"
    message: Gateway is accepted
    observedGeneration: 2
    reason: Accepted
    status: "True"
    type: Accepted
  - lastTransitionTime: "2022-12-16T19:15:41Z"
    message: Listeners are not valid
    observedGeneration: 2
    reason: ListenersNotValid
    status: "False"
    type: Programmed
  listeners:
  - attachedRoutes: 0
    conditions:
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Only one HTTP port is supported
      observedGeneration: 2
      reason: PortUnavailable
      status: "False"
      type: Accepted
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Invalid listener, see other listener conditions for details
      observedGeneration: 2
      reason: Invalid
      status: "False"
      type: Programmed
    name: alternate
    supportedKinds: []
  - attachedRoutes: 0
    conditions:
    - lastTransitionTime: "2022-12-16T19:15:41Z"
      message: Valid listener
      observedGeneration: 2
      reason: Programmed
      status: "True"
      type: Programmed
    name: http
    supportedKinds:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
```

The Gateway as a whole should still be `Accepted=true` since some part of the config is valid, per: https://gateway-api.sigs.k8s.io/geps/gep-1364/#accepted

observedGenerations are all correct, though the status/validity isn't all there